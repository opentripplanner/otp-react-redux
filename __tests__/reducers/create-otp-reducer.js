/* globals describe, expect, it */

import createOtpReducer from '../../lib/reducers/create-otp-reducer'

describe('reducers > createOtpReducer >', () => {
  describe('default reducer handlers >', () => {
    it('should return default initial state', () => {
      const defaultReducer = createOtpReducer({}, {})
      expect(
        stripDateAndTimeFromReducerState(
          defaultReducer(undefined, {type: 'init'})
        )
      ).toMatchSnapshot()
    })

    it('should handle PLAN_REQUEST action', () => {
      const defaultReducer = createOtpReducer({}, {})
      const defaultReducerState = defaultReducer(undefined, {type: 'init'})
      expect(
        stripDateAndTimeFromReducerState(
          defaultReducer(defaultReducerState, {type: 'PLAN_REQUEST'})
        )
      ).toMatchSnapshot()
    })

    it('should handle PLAN_ERROR action', () => {
      const defaultReducer = createOtpReducer({}, {})
      const defaultReducerState = defaultReducer(undefined, {type: 'init'})
      const pendingReducerState = defaultReducer(defaultReducerState, {type: 'PLAN_REQUEST'})
      expect(
        stripDateAndTimeFromReducerState(
          defaultReducer(pendingReducerState, {
            type: 'PLAN_ERROR',
            payload: new Error('An error')
          })
        )
      ).toMatchSnapshot()
    })

    it('should handle PLAN_RESPONSE action', () => {
      const defaultReducer = createOtpReducer({}, {})
      const defaultReducerState = defaultReducer(undefined, {type: 'init'})
      const pendingReducerState = defaultReducer(defaultReducerState, {type: 'PLAN_REQUEST'})
      expect(
        stripDateAndTimeFromReducerState(
          defaultReducer(pendingReducerState, {
            type: 'PLAN_RESPONSE',
            payload: { fakeResponse: true }
          })
        )
      ).toMatchSnapshot()
    })
  })

  describe('reducer with postProcessor >', () => {
    it('should post-process results', () => {
      const postProcessor = jest.fn((state, results) => ({ processed: true }))
      const defaultReducer = createOtpReducer({}, {}, postProcessor)
      const defaultReducerState = defaultReducer(undefined, {type: 'init'})
      const pendingReducerState = defaultReducer(defaultReducerState, {type: 'PLAN_REQUEST'})
      expect(
        stripDateAndTimeFromReducerState(
          defaultReducer(pendingReducerState, {
            type: 'PLAN_RESPONSE',
            payload: { fakeResponse: true }
          })
        )
      ).toMatchSnapshot()
      expect(postProcessor.mock.calls).toMatchSnapshot()
    })
  })
})

function stripDateAndTimeFromQuery (query) {
  // assert that date and time exist and then remove them
  expect(query.date).toEqual(expect.stringMatching(/\d{4}-\d{2}-\d{2}/))
  delete query.date
  expect(query.time).toEqual(expect.stringMatching(/\d{2}:\d{2}/))
  delete query.time
}

function stripDateAndTimeFromReducerState (reducerState) {
  stripDateAndTimeFromQuery(reducerState.currentQuery)

  // also remove from searches
  if (reducerState.searches) {
    reducerState.searches.forEach(search => {
      stripDateAndTimeFromQuery(search.query)
    })
  }

  return reducerState
}
