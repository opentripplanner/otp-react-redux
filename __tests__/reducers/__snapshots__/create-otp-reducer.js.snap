// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers > createOtpReducer > default reducer handlers > should handle PLAN_ERROR action 1`] = `
Object {
  "activeSearch": 0,
  "config": Object {
    "autoPlan": true,
    "debouncePlanTimeMs": 0,
  },
  "currentQuery": Object {
    "departArrive": "NOW",
    "from": null,
    "mode": "TRAM,BUS,GONDOLA,WALK",
    "to": null,
    "type": "ITINERARY",
  },
  "overlay": Object {
    "bikeRental": Object {
      "stations": Array [],
    },
  },
  "searches": Array [
    Object {
      "activeItinerary": 0,
      "activeLeg": null,
      "activeStep": null,
      "pending": false,
      "planResponse": Object {
        "error": [Error: An error],
      },
      "query": Object {
        "departArrive": "NOW",
        "from": null,
        "mode": "TRAM,BUS,GONDOLA,WALK",
        "to": null,
        "type": "ITINERARY",
      },
    },
  ],
  "ui": Object {
    "showExtendedSettings": false,
  },
}
`;

exports[`reducers > createOtpReducer > default reducer handlers > should handle PLAN_REQUEST action 1`] = `
Object {
  "activeSearch": 0,
  "config": Object {
    "autoPlan": true,
    "debouncePlanTimeMs": 0,
  },
  "currentQuery": Object {
    "departArrive": "NOW",
    "from": null,
    "mode": "TRAM,BUS,GONDOLA,WALK",
    "to": null,
    "type": "ITINERARY",
  },
  "overlay": Object {
    "bikeRental": Object {
      "stations": Array [],
    },
  },
  "searches": Array [
    Object {
      "activeItinerary": 0,
      "activeLeg": null,
      "activeStep": null,
      "pending": true,
      "planResponse": null,
      "query": Object {
        "departArrive": "NOW",
        "from": null,
        "mode": "TRAM,BUS,GONDOLA,WALK",
        "to": null,
        "type": "ITINERARY",
      },
    },
  ],
  "ui": Object {
    "showExtendedSettings": false,
  },
}
`;

exports[`reducers > createOtpReducer > default reducer handlers > should handle PLAN_RESPONSE action 1`] = `
Object {
  "activeSearch": 0,
  "config": Object {
    "autoPlan": true,
    "debouncePlanTimeMs": 0,
  },
  "currentQuery": Object {
    "departArrive": "NOW",
    "from": null,
    "mode": "TRAM,BUS,GONDOLA,WALK",
    "to": null,
    "type": "ITINERARY",
  },
  "overlay": Object {
    "bikeRental": Object {
      "stations": Array [],
    },
  },
  "searches": Array [
    Object {
      "activeItinerary": 0,
      "activeLeg": null,
      "activeStep": null,
      "pending": false,
      "planResponse": Object {
        "fakeResponse": true,
      },
      "postProcessedResults": null,
      "query": Object {
        "departArrive": "NOW",
        "from": null,
        "mode": "TRAM,BUS,GONDOLA,WALK",
        "to": null,
        "type": "ITINERARY",
      },
    },
  ],
  "ui": Object {
    "showExtendedSettings": false,
  },
}
`;

exports[`reducers > createOtpReducer > default reducer handlers > should return default initial state 1`] = `
Object {
  "activeSearch": 0,
  "config": Object {
    "autoPlan": true,
    "debouncePlanTimeMs": 0,
  },
  "currentQuery": Object {
    "departArrive": "NOW",
    "from": null,
    "mode": "TRAM,BUS,GONDOLA,WALK",
    "to": null,
    "type": "ITINERARY",
  },
  "overlay": Object {
    "bikeRental": Object {
      "stations": Array [],
    },
  },
  "searches": Array [],
  "ui": Object {
    "showExtendedSettings": false,
  },
}
`;

exports[`reducers > createOtpReducer > reducer with postProcessor > should post-process results 1`] = `
Object {
  "activeSearch": 0,
  "config": Object {
    "autoPlan": true,
    "debouncePlanTimeMs": 0,
  },
  "currentQuery": Object {
    "departArrive": "NOW",
    "from": null,
    "mode": "TRAM,BUS,GONDOLA,WALK",
    "to": null,
    "type": "ITINERARY",
  },
  "overlay": Object {
    "bikeRental": Object {
      "stations": Array [],
    },
  },
  "searches": Array [
    Object {
      "activeItinerary": 0,
      "activeLeg": null,
      "activeStep": null,
      "pending": false,
      "planResponse": Object {
        "fakeResponse": true,
      },
      "postProcessedResults": Object {
        "processed": true,
      },
      "query": Object {
        "departArrive": "NOW",
        "from": null,
        "mode": "TRAM,BUS,GONDOLA,WALK",
        "to": null,
        "type": "ITINERARY",
      },
    },
  ],
  "ui": Object {
    "showExtendedSettings": false,
  },
}
`;

exports[`reducers > createOtpReducer > reducer with postProcessor > should post-process results 2`] = `
Array [
  Array [
    Object {
      "activeSearch": 0,
      "config": Object {
        "autoPlan": true,
        "debouncePlanTimeMs": 0,
      },
      "currentQuery": Object {
        "departArrive": "NOW",
        "from": null,
        "mode": "TRAM,BUS,GONDOLA,WALK",
        "to": null,
        "type": "ITINERARY",
      },
      "overlay": Object {
        "bikeRental": Object {
          "stations": Array [],
        },
      },
      "searches": Array [
        Object {
          "activeItinerary": 0,
          "activeLeg": null,
          "activeStep": null,
          "pending": true,
          "planResponse": null,
          "query": Object {
            "departArrive": "NOW",
            "from": null,
            "mode": "TRAM,BUS,GONDOLA,WALK",
            "to": null,
            "type": "ITINERARY",
          },
        },
      ],
      "ui": Object {
        "showExtendedSettings": false,
      },
    },
    Object {
      "fakeResponse": true,
    },
  ],
]
`;
